### ISSUE TITLE
Align terminal output to the bottom of the window (as option, especially with _quake)

### ISSUE THREAD

**Issue by msacharewicz** [microsoft/terminal#14773]
# Description of the new feature/enhancement

I would like you to consider allowing users to align the output of the terminal to the bottom edge of the terminal window.

Currently, the terminal output is being rendered on the top edge of the terminal window. This is often just a single prompt line:

![obraz](https://user-images.githubusercontent.com/34579814/216163974-4089f1c9-7032-4d1a-91b2-536b1027bfb9.png)

While working with terminal, though, the screen gets filled with output and prompt usually ends up on the bottom for the entirety of the session:

![obraz](https://user-images.githubusercontent.com/34579814/216164203-919f5bf3-88af-48fb-a88b-62c6433b315b.png)

Allowing user to align the output to the bottom of the screen would keep the prompt consistently in the same place, giving less distracting and more visually pleasing experience (subjectively):

![obraz](https://user-images.githubusercontent.com/34579814/216164336-5fb23577-e677-44f5-b7f8-d1924fef832b.png)

It would also be a more natural experience, when using the console quake-style.

I am well aware, that the same effect can be achieved in most shells using various tricks, but:
- they mostly depend on filling the output with missing lines, which is slow,
- they will break once the terminal window gets resized.

As this is my first appearance here, HUGE THANKS for the awesome work done already! ❤😊

**Comment by zadjii-msft** [comment]
Okay so re-reading theis more closely - this isn't "upside down mode" (where newlines go up instead of down)[^1]. 

This could be done with a new setting (profile? global?) for enabling the terminal to pre-fill the buffer with some newlines. We could use `--inheritCursor` then, when starting the connection.


[1]: This was a dumb hackathon project I tried once, but was ultimately way too crazy to even consider shipping.

**Comment by msacharewicz** [comment]
@zadjii-msft Nope, not the "upside down mode". And I actually doubt that _pre-filling the buffer_ is the right approach.

Just trim the output buffer rect to the rows received by the buffer and then align such rect to the bottom edge of the parent rect, if there are less rows than the window size:

![obraz](https://user-images.githubusercontent.com/34579814/216333690-e0bc1a7b-7cea-434a-ac3c-4343d6d711de.png)

Mind, that any empty/whitespace rows that _are_ actually being sent to the buffer by some "full-screen" apps (as with, let's say, the basic `top` command output) _should_ be included inside such rect.

**Comment by zadjii-msft** [comment]
So, inserting fake newlines would have been an implementation detail, to avoid the scenario where the viewport would have a top that's <0. I'm sure there's a huge swath of code that assumes a `Viewport` can't have a negative top member. 

I guess that wouldn't work for a resize though - as soon as you add lines, the prompt would no longer be at the bottom. So maybe a negative viewport top is the only way to do that. Presumably, the renderer could just assume that any rows of text at y<0 would be filled with blank spaces.


The other hard problem would be how this gets enabled in the settings. It seems kinda like that would make sense as a per-profile option (since this is something that controls the behavior of the control). But at the same time, it also seems like something that only makes sense with a global `snapToSide: top/bottom` property, so maybe this is a global "I want to snap the viewport to the bottom of the window" setting that controls all your terminals. 

There's clearly lots for someone to noodle on here.

**Comment by msacharewicz** [comment]
Well, hard for me to argue, since even though I tried to give a look at the code, it's hard to get around with just some basic, extremely dusty cpp knowledge.

All I can say (by playing with the current public release of the Terminal) is that:
1. there is some kind of a container/rect that holds the rendered content of the buffer;
2. this container can be larger than the viewport (thus, scrolling works);
3. when the buffer is larger than viewport size, the container aligns to bottom by default,
4. when the buffer is smaller than viewport size, it is aligned to the top of the viewport,
5. when I resize the Terminal window by dragging the bottom border up:
    a. until I reach the actual buffer content, it does not move up;
    b. when I reach it, it starts to move up along the border.

Therefore, I hope I assume it correctly, that:
- the buffer container/rect is actually already trimmed to the size of the received buffer content, not of the full viewport size, otherwise when resizing the window (ref. 4a) it would go up;
- that it is freely alignable within the viewport (ref. 3).

Thus, unless the above behaviour is achieved by some extensive, low-level hacks, it seems that all that's needed is a change of alignment when the buffer container/rect is smaller than the viewport.

But, of course, all these assumptions come from some damn old and barely remembered experience I had with WINAPI some 20 years ago. I may be babbling, if so, forgive me 😊

As for the option scopes, having it as both a profile and global (profile-default) choice would be a charm.

**Comment by 237dmitry** [comment]
Could you say what terminal emulator has such functionality?

**Comment by msacharewicz** [comment]
@237dmitry: not sure, what you ask for specifically and how it relates, but I'll do my best.

- **Which terminal exhibits behavior which I did describe in the previous comment, points 1-5?** 
   *Windows Terminal*. This very app.

- **Which other, existing terminal emulator has the feature that I described in this suggestion?**
    None that I know of, sadly. The suggestion is based on:
    - the behavior of few (very) old terminal displays that mimicked the early print terminals ("paper goes always up"),
    - the perceived gains (ie. when prompt stays in same place, it's easier to focus on),
    - the popular demand (ie. the google results on _"how to align the prompt to bottom in (bash|powershell|..."_),
    - the inefficiency of in-shell workarounds.


To simplify my last comment, lets use the picture.

- There is already _some_ rect that grows with buffer content. It sticks to the bottom edge whenever you drag that edge it up.
- Suggested feature: allow this rect to always stick to the bottom edge.

** **

![obraz](https://user-images.githubusercontent.com/34579814/216581628-6df5ba62-5387-4265-8c4b-37fdf41e71fc.png)

**Comment by o-sdn-o** [comment]
It seems to me that such a layout breaks caret positioning for applications that use it on a scrollbuffer, see vt-sequence CUP `\e[ Y ; X H`. Shifting all text up after output is not compatible with caret positioning. Moreover, after clearing the viewport, the caret should always be positioned at position 1,1 (apparently there is some convention), respectively, the shell draws the prompt at coordinate 1,1.

**Comment by o-sdn-o** [comment]
As a workaround, you can add CUP `\033[999H` to the beginning of the colored prompt in `~/.bashrc`

```bash
if [ "$color_prompt" = yes ]; then                                                                  
    PS1='${debian_chroot:+($debian_chroot)}\[\033[999H\033[01;32m\]\u@\h\[\033[00m\]:\[\033[01;34m\] blablabla
else                                                                                                
    PS1='${debian_chroot:+($debian_chroot)}\u@\h:\w\$ '                                             
fi
```

This way the shell prompt is always at the bottom as you like.

**Comment by zadjii-msft** [comment]
This would definitely be a virtual kind of mode. In the absolutely initial case, with 1 row output, that the real, "mutable" viewport would only have one visible row. So moving the cursor down from there would just "reveal" that many lines. It'd just be like, a renderer transform or something silly like that. 

oh actually that's a really good idea. That might make this easier.

**Comment by o-sdn-o** [comment]
You can use `\033[999;999H\n` instead of `\033[999H` to avoid overriding the viewport contents (this adds an empty line at the viewport bottom).

**Comment by o-sdn-o** [comment]
For the powershell case

```powershell
function prompt  
{  
    $E=[char]27
    "$E[999;999H $E[32mPS $($executionContext.SessionState.Path.CurrentLocation)> $E[0m"  
}
```

**Comment by o-sdn-o** [comment]
> This would definitely be a virtual kind of mode. In the absolutely initial case, with 1 row output, that the real, "mutable" viewport would only have one visible row. So moving the cursor down from there would just "reveal" that many lines. It'd just be like, a renderer transform or something silly like that.

This kind of mode can't survive after `Erase viewport` `\e[2J` after filling the viewport with blank lines.
